pytorch_translate/beam_decode.py

```diff
@@ -517,9 +530,95 @@ def _encode(self, encoder_input, reorder_indices):
             encoder_outs.append(encoder_out)
         return encoder_outs, incremental_states
 
-    def _decode(self, tokens, encoder_outs, incremental_states):
+    @staticmethod
+    def gather_probs(all_translation_tokens: List[Optional[Any]], all_probs: List[Any]):
+        """
+        Maps probabilities for multiple models with different output softmax
+        dimensions to the same combined token space.
+
+        Model 1:
+        possible_translation_tokens: [3, 7, 8, 9]
+        probs: [0.25, 0.25, 0.25, 0.25]
+
+        Model 2:
+        possible_translation_tokens: [0, 3, 5]
+        probs: [0.4, 0.5, 0.1]
+
+        all_translation_tokens: [[3, 7, 8, 9], [0, 3, 5]]
+        all_probs: [[0.25, 0.25, 0.25, 0.25], [0.4, 0.5, 0.1]]
+        possible_translation_tokens = [0, 3, 5, 7, 8, 9] (order varies)
+        mapped_probs for model 1: [0  , 0.25, 0  , 0.25, 0.25, 0.25]
+        mapped_probs for model 2: [0.4, 0.5 , 0.1, 0   , 0   , 0]
+
+        avg_probs = [0.4, 0.75, 0.1, 0.25, 0.25, 0.25] (order varies but
```

Comment:
This is the sum, not the average.

Reply:
Could you clarify in doc whether this is summing log prob?

Reply:
In practice the values are log probs, and they are just summed, not averaged. This should certainly be clearer in the code (misleading variable names don't help), but that's mostly unrelated to this change.